### PR TITLE
Add filtering of huge tag lists to get under the nginx limit

### PR DESCRIPTION
**Pull request by achbed**
Adds support for automagically trimming large lists of surrogate keys that would break nginx and varnish. This will attempt to trim keys by category from largest to smallest until it fits within the defined header size. 

To do this, it defines a category as everything up to and including the final dash in the key. For each category we trim, we use category-huge as the key instead of all the component keys.

Support for clearing -huge keys for the appropriate categories is also added to the purge functions.

### CODE REVIEW

**Comment by danielbachhuber** [comment]
Thanks @achbed! Can you include some tests for this too?

**Comment by achbed** [comment]
I'm done making changes to the PR (as long as this passes).

**Comment by danielbachhuber** [comment]
> I'm done making changes to the PR (as long as this passes).

Cool, looks pretty good on my end. I'll follow up when I've had the chance to run it past Josh.

File: inc/class-emitter.php [hunk 1]
```diff
@@ -41,6 +41,13 @@ class Emitter {
 	const HEADER_KEY = 'Surrogate-Key';
 
 	/**
+	 * Maximum header length.
+	 *
+	 * @var integer
+	 */
+	const HEADER_MAX_LENGTH = 4096;
```
Comment (danielbachhuber) [inline inc/class-emitter.php]: Why did you choose this size? The real-world constraint is something close to 32kb

Comment (achbed) [inline inc/class-emitter.php]: It was easy to test. I've upped that limit in the next PR update (and fixed a typo that caused the unit tests to fail).

**Comment by danielbachhuber** [comment]
Thanks again for your work on this, @achbed.

**Pull request merged**